
## Hypothesis Testing On Light Theme vs Dark Theme 

1. Understanding the data

In [1]:
import pandas as pd 
from scipy.stats import ttest_ind 
df = pd.read_csv("Data/website_ab_test.csv")
print(df.head())

         Theme  Click Through Rate  Conversion Rate  Bounce Rate  \
0  Light Theme            0.054920         0.282367     0.405085   
1  Light Theme            0.113932         0.032973     0.732759   
2   Dark Theme            0.323352         0.178763     0.296543   
3  Light Theme            0.485836         0.325225     0.245001   
4  Light Theme            0.034783         0.196766     0.765100   

   Scroll_Depth  Age   Location  Session_Duration Purchases Added_to_Cart  
0     72.489458   25    Chennai              1535        No           Yes  
1     61.858568   19       Pune               303        No           Yes  
2     45.737376   47    Chennai               563       Yes           Yes  
3     76.305298   58       Pune               385       Yes            No  
4     48.927407   25  New Delhi              1437        No            No  


So given the website performance dataset, we need to understand how the theme of the website is affecting the user behavior.

Next we need to understand the dataset overall by finding the summary statistics of dataset.

In [ ]:
data_summary= {'Number of Records': df.shape[0],
               'Columns': df.shape[1],
               'Missing values': df.isnull().sum(),
               'Numerical columns summary': df.describe()}

In [4]:
print(data_summary)

{'Number of Records': 1000, 'Columns': 10, 'Missing values': Theme                 0
Click Through Rate    0
Conversion Rate       0
Bounce Rate           0
Scroll_Depth          0
Age                   0
Location              0
Session_Duration      0
Purchases             0
Added_to_Cart         0
dtype: int64, 'Numerical columns summary':        Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
count         1000.000000      1000.000000  1000.000000   1000.000000   
mean             0.256048         0.253312     0.505758     50.319494   
std              0.139265         0.139092     0.172195     16.895269   
min              0.010767         0.010881     0.200720     20.011738   
25%              0.140794         0.131564     0.353609     35.655167   
50%              0.253715         0.252823     0.514049     51.130712   
75%              0.370674         0.373040     0.648557     64.666258   
max              0.499989         0.498916     0.799658     79.997108   

The above output explains that the dataset contains 1000 records across 10 columns and it has no missing values.

The numerical column summary describes that the columns:

-Click Through Rate: values ranges from 0.01 to 0.50 with a mean of 0.26

-Conversion Rate: values range from 0.01 to 0.50 with a mean of 0.25.

-Bounce Rate: values range from 0.20 to 0.80 with a mean of 0.51.

-Scroll Depth: values range from 20.01 to 80 with a mean of 50.32.

-Age: values range from 18 to 65 years with a mean of about 41.5 years.

-Session Duration: this columly widely aries from 38 seconds to nearly 1800 seconds( 30 min) with a mean session dusation of 925 seconds(15 min).

Since we have a better undertsanding of our data let's analyze the performance between both themes based on the averages of all metrics for each theme.

In [6]:
df_numeric = df.select_dtypes(include=["number"])  # Select only numeric columns
theme_performance = df.groupby("Theme")[df_numeric.columns].mean()

theme_performance_sorted = theme_performance.sort_values(by="Conversion Rate", ascending=False)
print(theme_performance_sorted)

             Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
Theme                                                                         
Light Theme            0.247109         0.255459     0.499035     50.735232   
Dark Theme             0.264501         0.251282     0.512115     49.926404   

                   Age  Session_Duration  
Theme                                     
Light Theme  41.734568        930.833333  
Dark Theme   41.332685        919.482490  


This output signifies that the click through rate for dark theme is slightly higher than light theme.

The conversion rate is marginally higher for light theme. 

The bounce rate is slightly higher for dark theme. 

User scroll slightly further on average of 50.74% on light theme than dark theme which is 49.93%. 

The average age of users for light theme is 41.73 while 41.33 for dark users.

The average session duration is slightly longer for users on light theme(930.83 sec) compared to dark theme(919.48 sec).

This shows that light theme slightly outperforms dark theme in the given metrics.

# Hypothesis Testing 

For conversion rates:

Null Hypothesis(H0)- There is no difference in conversion rates between light and dark theme.

Alternate Hypothesis(H1)- There is difference in conversion rates between light and dark theme.

In [12]:
# 2-sample t test to compare means of two independent samples
conversion_rates_light= df[df['Theme']== 'Light Theme']['Conversion Rate']
conversion_rates_dark= df[df['Theme']=='Dark Theme']['Conversion Rate']

t_state_conversion, p_state_conversion= ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var= False)

print(t_state_conversion, p_state_conversion)                            

0.4748494462782632 0.6349982678451778


t-test- tells us difference between sample means relative to variation within samples

p-state- tells us whether the difference between the means is statistically significant (p_state<0.05 we reject null hypothesus which means are significantly different, p_state >0.05 we fail to reject null hypothesus that is means are significantly different)


The result from above two-sample t-test gives a p value approx 0.635 . p_state > 0.05 says we don't have enough evidence to reject null hypothesis. Which concludes the conversion rates between light theme and dark theme is statistically different.

In [19]:
#Click-Through-Rate

click_through_rates_light= df[df['Theme']== 'Light Theme']['Click Through Rate']
click_through_rates_dark= df[df['Theme']=='Dark Theme']['Click Through Rate']

t_state_ctr, p_state_ctr= ttest_ind(click_through_rates_light, click_through_rates_dark, equal_var= False)

print(t_state_ctr,
      p_state_ctr)

-1.9781708664172253 0.04818435371010704


The two-sample t-test for the CTR between the light theme and dark theme yields a p-value of approximately 0.048. This p value is slightly below our significance level of 0.05, indicating there is ststistically no difference in CTR between Light and Dark theme.

In [14]:
#Bounce rate

bounce_rates_light = df[df['Theme']== 'Light Theme']['Bounce Rate']
bounce_rates_dark= df[df['Theme']=='Dark Theme']['Bounce Rate']

t_state_bounce, p_state_bounce= ttest_ind(bounce_rates_light, bounce_rates_dark, equal_var= False)

print(t_state_bounce, p_state_bounce)

-1.2018883310494073 0.229692077505148


In [16]:
#scroll depth

scroll_depth_light= df[df['Theme']== 'Light Theme']['Scroll_Depth']
scroll_depth_dark= df[df['Theme']=='Dark Theme']['Scroll_Depth']

t_state_scroll, p_state_scroll= ttest_ind(scroll_depth_light, scroll_depth_dark, equal_var= False)

print(t_state_scroll, p_state_scroll)

0.7562277864140986 0.4496919249484911


In [17]:
comparison_table = pd.DataFrame({
    'Metric':['Click Through Rate', 'Conversion Rate','Bounce Rate', 'Scroll Depth'],
    'T-Statistic': [t_state_ctr, t_state_conversion,t_state_bounce,t_state_scroll],
    'P-Value': [p_state_ctr,p_state_conversion,p_state_bounce,p_state_scroll]
})
comparison_table

Metric  T-Statistic   P-Value
0  Click Through Rate    -1.978171  0.048184
1     Conversion Rate     0.474849  0.634998
2         Bounce Rate    -1.201888  0.229692
3        Scroll Depth     0.756228  0.449692

Click Through Rate : the test reveals a statistically signaficant difference, with the Dark Theme likely performing better(p-value= 0.048).

Conversion Rate: No statistically significant difference was found(p-value= 0.635).

Bounce Rate: There's no statistically significant difference in bounce rates between the themes(p-value= 0.230)

Scroll Depth: Similarly, no statistically significant difference is observed in scroll depth(p-vale= 0.450).

## CONCLUSION

While the two themes peroform similar across most metric, the Dark theme has a slight edge in terms of engaging users to click through while other metrics does not significantly affect user behavior according to the data provided.